### ISSUE TITLE
Incorrect flipping of door textures compared with open/closed state

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest_game#1361]
I noticed that something odd is going on with door/trapdoor textures. Whenever you open or close them, the textures may get flipped incorrectly.
Just look at the following screenshots:

 (**FIXED**) Closed trapdoors with equal facedir stacked above each other. The texture is flipped on the opposite side.
![](https://i.imgur.com/nzXyqIF.png)

(**FIXED**) Two trapdoors, same facedir, one closed, one opened. Again, the texture is flipped incorrectly for the open one:
![](https://i.imgur.com/PfeDVjR.png)

(**FIXED**) The sides of the trapdoors get flipped when opening, too: Note the yellow/blue stripe should be at the right for the open trapdoor (However, I am not sure about that one. Maybe this works when I change the “side” texture):
![](https://i.imgur.com/B7d4XJQ.png)

(**LOW PRIORITY, NON-TRIVIAL**) For doors, the inner and outer “sides” get flipped, too, when opening them:
![](https://i.imgur.com/KAMFXUG.png)
![](https://i.imgur.com/yH36uAW.png)

I am not sure if these are all issues. Try the test textures for yourselves and test a bit with the doors and trap doors.

Here you can use the test textures I used:
![test_trapdoor_topbottom](https://cloud.githubusercontent.com/assets/1675853/19893603/0616c140-a04a-11e6-84ef-fb0b26a450b3.png)
![test_trapdoor_side](https://cloud.githubusercontent.com/assets/1675853/19893604/061aaf76-a04a-11e6-99a2-ae1a5d792767.png)
![test_door](https://cloud.githubusercontent.com/assets/1675853/19893627/1e71f598-a04a-11e6-81e5-5e73a4f24000.png)

I have added several “markers” on the test textures to make them easily distinguishable. The dotted line on the front textures mark the “hinge”. The colors on the sides are for orientation. The yellow square on the door marks the top left of the door image (not the physical “left”). The letters on the trapdoor mean “up”, “down”, “left” and “right”.

**Comment by sofar** [comment]
This is probably fixable for doors, since those are mesh nodes now.

The problem with trapdoors is that they're nodeboxes. I need to see if we can play with transforms to get this right :/

**Comment by Wuzzy2** [comment]
Can you map different textures on 2 opposite sides (e.g. top and bottom) of a nodebox? If yes, it should be fixable. If not, then we are indeed in trouble. xD

**Comment by sofar** [comment]
The open trapdoor is already using per-side tiles, we can transformfx them to the right angle.

**Comment by Thomas--S** [comment]
I think this was fixed with #1202.
@Wuzzy2 usually uses the stable version, where this  fix isn't included yet.

**Comment by paramat** [comment]
See #1202 perhaps Wuzzy2's more detailed test textures are showing a bug missed with simpler textures?

**Comment by Wuzzy2** [comment]
Oops, sorry. I wasn't using the dev version this time, I guess. Usually I use the dev version when bug hunting but this time I have just forgotten it. So sorry. :-(

Trapdoor top and bottom images are OK now. :-)

The following problems remain:
- Sides of doors flipped when opening/closing (see above)
- Sides of trapdoors flipped when opening/closing (see above)

And finally, there is a bug so obvious we all have overlooked it in plain sight:
Notice the door front turns from I to II when opening/closing? This should not be the case, the door texture should stay the same.
What this means is, you cannot currently use two different textured door sides because the texture will change whenever you open or close the door.

This bug has of course been invisible because texture drawers usually just worked around this (intentional or not) by using basically the same texture on both sides, just flipped on the other side.

**Comment by sofar** [comment]
Since the pictures are of the old version, they show both old bugs and new bugs. I think it would be best to make a few new screenshots since I can't seem to reproduce the trapdoor side issue myself (it all looked ok to me when I tried yesterday, but maybe I'm not seeing it right).

**Comment by sofar** [comment]
The issue with doors is that there's essentially no "inside" or "outside", since it would require more nodes make the textures all entirely consistent. There's essentially only a left-hinged door and a right-hinged door. When you open a left-hinged door, the node is replaced with a right-hinged door (and rotated 90 degrees). When you close it again, it's again replaced with  a left-hinged door (and rotated back).

Because of this, the "inside" of the door is essentially flipped every time.

**Comment by paramat** [comment]
In that case i think doors are ok as they are, the simplicity is more important than the ability to use different textures for the sides, which would be very rarey done anyway.
If it's the same issue for trapdoor i have the same opinion.

**Comment by Ferk** [comment]
@sofar What about making the left-hinged door have the outside in the side in which the right-hinged door has the inside?

While this still wouldn't really allow you to make a texture that is clearly the outside / inside, at least it would be possible to make each side different without having the texture switching in an unexpected way when the door opens/closes.

```
1)  (front of door(outside) is on the bottom)
....
....
....
hvvv

2)   (front of door(outside) is on the bottom!)
....
....
....
^^^h

3)   (front of door(outside) is on the inside!)
>...
>...
>...
h...
```

**Comment by sofar** [comment]
@ferk what you drew is the existing nodes (it's 2 nodes, you can go from your 1 to 4 by rotation, and from your 2 to 3)

Perhaps there's a UV map mirror issue here where I need to flip a texture or something - I'll need to look in blender for that.

**Comment by Ferk** [comment]
@sofar yes, I know there are those two nodes, I actually meant changing the UV map use different texture for each side (that's why I wrote `^^^h` instead of the `vvvh` you wrote for the right-hinged door)

**Comment by Wuzzy2** [comment]
Note: I just updated the issue and marked the fixed parts of this issue.

**Comment by sofar** [comment]
@ferk it is using a different texture for each side.

https://github.com/minetest/minetest_game/blob/master/mods/doors/textures/doors_door_wood.png

**Comment by Ferk** [comment]
@sofar What I mean is that both left-hinged and right-hinged use the same texture in the same side:

Both of them are (I)-front (II)-back. If one of them was instead (II)-front (I)-back (ie. the outside face of one mesh had the inside face of the other), then the opened door would be in the right orientation as I was trying to explain in the previous comments.

FRONT:
![](http://i.imgur.com/y4Q0sI3.png)
BACK:
![](http://i.imgur.com/rzoRKzl.png)

The texture is already mirrored, what I mean is that you can make the rotation work fine if instead of mirroring you switched the UV mapping, using different faces. As you can see the PNG you linked, you already have the texture mirrored in there anyway.

Maybe the current UV mapping was done intentionally like this to have double doors that use the same texture, mirrored... but then a decision has to be made: either double doors with same texture or consistent rotation for inside-outside textures.

**Comment by sofar** [comment]
Yes, the idea was to use 1 texture for all models.

I need to load up the model and UV map in blender to see if I haven't made any mistakes, which there seem to be according to the screenshots. But, I've got other things I'm working on as well, and this isn't necessarily high priority (it's a visual problem, and it took 5 months to get spotted).

**Comment by sofar** [comment]
Having spent some time looking at the issue again, I think I want to leave it as is. It exposes a painful miscalculation on my side - we should have either gone for a simpler UV map or with 4 nodes.

I have no solution and I don't feel like making one. Perhaps once we have meta set nodedef this could be solved easily, or we redo doors after 0.4.16 and fix this issue with 4 nodes instead of 2.

**Comment by b3u** [comment]
Not to be rude, but Is there a reason that more than a year later, this is still an open issue? Is it due to lack of volunteers, or low priority, etc.?

**Comment by paramat** [comment]
b3u the usual issue, lack of core dev time, also it is low priority.

I'm fine with leaving doors as they are, for now at least, it just may mean some care when creating textures for them.

The remaining problem is:
> Sides of trapdoors flipped when opening/closing (see above)

Wuzzy2's test texture should be used to recheck the bug, then it may be possible to fix this easily using texture transforms.

**Comment by b3u** [comment]
I still have the issue with *normal* doors. It's not clear to me how to fix this. It means that the back of my door becomes the front when open. My door texture is that of a picture frame.

**Comment by paramat** [comment]
See https://github.com/minetest/minetest_game/issues/1361#issuecomment-298146842 non-trivial to fix. Best for now to have matching textures.

**Comment by paramat** [comment]
> The sides of the trapdoors get flipped when opening, too: Note the yellow/blue stripe should be at the right for the open trapdoor (However, I am not sure about that one. Maybe this works when I change the “side” texture):

After much experimenting, it seems the 'side' texture has to be of the following form:
![doors_trapdoor_side_new128](https://user-images.githubusercontent.com/3686677/57112277-3ddc7e00-6d37-11e9-9a1e-4e622a2216a9.png)
The lowest 2 rows are the sides when closed, the highest 2 rows are the sides when open and this is drawn upright , not inverted.
Wuzzy2's test side texture didn't work because the top 2 rows were inverted.
The texture transforms are altered to:
```
	def_opened.tiles = {
			def.tile_side,
			def.tile_side .. '^[transform2',
			def.tile_side .. '^[transform3',
			def.tile_side .. '^[transform1',
			def.tile_front .. '^[transform46',
			def.tile_front .. '^[transform6'
	}
```
![screenshot_20190502_230456](https://user-images.githubusercontent.com/3686677/57112390-b5aaa880-6d37-11e9-85c2-074e67afa704.png)

![screenshot_20190502_230511](https://user-images.githubusercontent.com/3686677/57112395-b9d6c600-6d37-11e9-819b-fca3315b758f.png)

![screenshot_20190502_230545](https://user-images.githubusercontent.com/3686677/57112399-bd6a4d00-6d37-11e9-8694-f84702a97b9b.png)

![screenshot_20190502_230551](https://user-images.githubusercontent.com/3686677/57112402-c0653d80-6d37-11e9-9496-a5f78c3821d3.png)

The side textures in the doors mod would need altering if we make this change.
WIll test this further.

**Comment by paramat** [comment]
Testing MTG master, both trapdoors flip their side textures when opened:

![screenshot_20190503_002031](https://user-images.githubusercontent.com/3686677/57112865-db38b180-6d39-11e9-96b2-fc1cd389426e.png)

![screenshot_20190503_002042](https://user-images.githubusercontent.com/3686677/57112870-df64cf00-6d39-11e9-9236-9ae1fd219cdd.png)

![screenshot_20190503_002117](https://user-images.githubusercontent.com/3686677/57112872-e25fbf80-6d39-11e9-8038-8a60fa096f2c.png)

![screenshot_20190503_002129](https://user-images.githubusercontent.com/3686677/57112874-e5f34680-6d39-11e9-80ea-77a9f516d840.png)

So the side textures do need editing anyway.

**Comment by paramat** [comment]
I can make a PR for new side textures and new tiles definition, it won't break any mods as the same pixel rows are used.

**Comment by b3u** [comment]
Much attention has been given to trapdoors. Remember that normal doors are still an issue.

**Comment by paramat** [comment]
b3u, we know.
You seem to have ignored https://github.com/minetest/minetest_game/issues/1361#issuecomment-488143372 and https://github.com/minetest/minetest_game/issues/1361#issuecomment-488373348
Fixing this is difficult because all doors in worlds would need automatically replacing with the new system.

Please stop ignoring what has been written, repeating the obvious and repeatedly complaining, it won't make a fix happen any sooner. You're being irritating.
You've made an unnecessary fuss in another issue too.

**Comment by paramat** [comment]
PR #2361

**Comment by sofar** [comment]
I think the only real way to fix it is to use 2 mesh nodes and do proper 
UV wrapping based on 2 separate 16x16 textures (top and side). This 
would be the easiest solution and would correctly work with rotating the 
trapdoor without any special magic. This would ultimately also work for 
the normal doors.

**Comment by b3u** [comment]
Paramat: I apologize for being irritating. I didn't intend to cause a fuss. I know I can be obsessive about things. I'm sorry.

**Comment by paramat** [comment]
Ok =) sorry i got annoyed.

**Comment by paramat** [comment]
sofar, i have a fix for trapdoors. After the hell we went through moving to meshnode doors i disapprove of doing the same for trapdoors unless it becomes essential to do so.

**Comment by paramat** [comment]
#2361 merged.

**Comment by sofar** [comment]
> sofar, i have a simple fix for trapdoors. After the hell we went through moving to meshnode doors i disapprove of doing the same for trapdoors unless it becomes essential to do so. Let's keep them simple if possible.

I think you may have broken texture packs with that change. It's probably not a big deal.

**Comment by An0n3m0us** [comment]
Why is this issue still open; the trapdoor sides seem to be fixed?

**Comment by paramat** [comment]
See first post, this issue is now only for the door issue.

**Comment by An0n3m0us** [comment]
Closed by #2372 or no?
